sandbox/plugins/cryptodirectory/src/main/java/org/opensearch/index/store/CryptoDirectory.java

```diff
@@ -0,0 +1,536 @@
+ * not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+/*
+ * Modifications Copyright OpenSearch Contributors. See
+ * GitHub history for details.
+ */
+
+package org.opensearch.index.store;
+
+import org.apache.lucene.store.BufferedIndexInput;
+import org.apache.lucene.store.IOContext;
+import org.apache.lucene.store.IndexInput;
+import org.apache.lucene.store.IndexOutput;
+import org.apache.lucene.store.LockFactory;
+import org.apache.lucene.store.NIOFSDirectory;
+import org.apache.lucene.store.OutputStreamIndexOutput;
+import org.opensearch.common.Randomness;
+import org.opensearch.common.SuppressForbidden;
+import org.opensearch.common.crypto.DataKeyPair;
+import org.opensearch.common.crypto.MasterKeyProvider;
+import org.opensearch.common.util.io.IOUtils;
+
+import javax.crypto.BadPaddingException;
+import javax.crypto.Cipher;
+import javax.crypto.IllegalBlockSizeException;
+import javax.crypto.NoSuchPaddingException;
+import javax.crypto.ShortBufferException;
+import javax.crypto.spec.IvParameterSpec;
+import javax.crypto.spec.SecretKeySpec;
+
+import java.io.EOFException;
+import java.io.FilterOutputStream;
+import java.io.IOException;
+import java.io.InputStream;
+import java.io.OutputStream;
+import java.nio.ByteBuffer;
+import java.nio.channels.Channels;
+import java.nio.channels.FileChannel;
+import java.nio.file.Files;
+import java.nio.file.Path;
+import java.nio.file.StandardOpenOption;
+import java.security.InvalidAlgorithmParameterException;
+import java.security.InvalidKeyException;
+import java.security.Key;
+import java.security.NoSuchAlgorithmException;
+import java.security.Provider;
+import java.security.SecureRandom;
+import java.util.Arrays;
+import java.util.Base64;
+import java.util.Map;
+import java.util.Optional;
+import java.util.concurrent.ConcurrentSkipListMap;
+import java.util.concurrent.atomic.AtomicLong;
+
+/**
+ * A hybrid directory implementation that encrypts files
+ * to be stored based on a user supplied key
+ *
+ * @opensearch.internal
+ */
+public final class CryptoDirectory extends NIOFSDirectory {
+    private Path location;
+    private Key dataKey;
+    private ConcurrentSkipListMap<String, String> ivMap;
+    private final Provider provider;
+
+    private final AtomicLong nextTempFileCounter = new AtomicLong();
+
+    CryptoDirectory(LockFactory lockFactory, Path location, Provider provider, MasterKeyProvider keyProvider) throws IOException {
+        super(location, lockFactory);
+        this.location = location;
+        ivMap = new ConcurrentSkipListMap<>();
```

Comment:
i wonder ```ivMap``` should be a property of CryptoDirectoryFactory so that it can be reused for hybridfs when using niofs and mmapfs.

Reply:
ofc, we will need to implement mamapfs (which can come later); but i think we need to think of the abstraction now.